packages/@lwc/engine-dom/src/patches/global-registry.ts

```diff
@@ -0,0 +1,386 @@
+function internalUpgrade(
+    instance: HTMLElement,
+    registeredDefinition: Definition,
+    instancedDefinition: Definition
+) {
+    setPrototypeOf(instance, instancedDefinition.UserCtor.prototype);
+    definitionForElement.set(instance, instancedDefinition);
+    // attributes patches when needed
+    if (instancedDefinition !== registeredDefinition) {
+        patchAttributes(instance, registeredDefinition, instancedDefinition);
+    }
+    // Tricking the construction path to believe that a new instance is being created,
+    // that way it will execute the super initialization mechanism but the HTMLElement
+    // constructor will reuse the instance by returning the upgradingInstance.
+    // This is by far the most important piece of the puzzle
+    upgradingInstance = instance;
+    new instancedDefinition.UserCtor();
+
+    const { observedAttributes, attributeChangedCallback } = instancedDefinition;
+    if (observedAttributes.size === 0 || isUndefined(attributeChangedCallback)) {
+        return;
+    }
+    const offset = getObservedAttributesOffset(registeredDefinition, instancedDefinition);
+    if (offset.size === 0) {
+        return;
+    }
+    // Approximate observedAttributes from the user class, but only for the offset attributes
+    offset.forEach((name) => {
+        if (nativeHasAttribute.call(instance, name)) {
+            const newValue = nativeGetAttribute.call(instance, name);
+            attributeChangedCallback!.call(instance, name, null, newValue);
+        }
+    });
+}
+
+function getDefinitionForConstructor(constructor: CustomElementConstructor): Definition {
+    if (!constructor || !constructor.prototype || typeof constructor.prototype !== 'object') {
+        throw new TypeError(`The referenced constructor is not a constructor.`);
+    }
+    let definition = definitionForConstructor.get(constructor);
+    if (isUndefined(definition)) {
+        definition = createDefinitionRecord(constructor);
+        definitionForConstructor.set(constructor, definition);
+    }
+    return definition;
+}
+
+export function patchCustomElementRegistry() {
+    const { customElements: nativeRegistry } = window;
+    const { define: nativeDefine, whenDefined: nativeWhenDefined, get: nativeGet } = nativeRegistry;
+
+    // patch for the global registry define mechanism
+    CustomElementRegistry.prototype.define = function define(
+        this: CustomElementRegistry,
+        tagName: string,
+        constructor: CustomElementConstructor,
+        options?: ElementDefinitionOptions
+    ): void {
+        if (options && options.extends) {
+            throw new DOMException(
+                'NotSupportedError: "extends" key in customElements.define() options is not supported.'
+            );
+        }
+        nativeGet.call(this, tagName); // SyntaxError if The provided name is not a valid custom element name.
+        tagName = tagName.toLowerCase();
```

Comment:
I think this line is not needed since the line before will throw for invalid names.